README.md
# Flask-Avatars
All avatar generators in one place:

* Default avatar
* Gravatar
* Robohash
* Social media avatar
* Identicon generation
* Avatar crop

## Links

* Documentation: https://flask-avatars.readthedocs.io/en/latest/
* PyPI: https://pypi.org/project/Flask-Avatars/

## License

This project is licensed under the MIT License (see the
`LICENSE` file for details).


CHANGES.rst
ChangeLog
---------

0.2.3
~~~~~

Release date: 2023/11/28

- Fix the compatibility issue with Flask ``3.x``.


0.2.2
~~~~~

Release date: 2019/3/4

- Fix ``TypeError`` when crop avatar without upload the raw image (use default avatar).


0.2.1
~~~~~

Release date: 2018/8/10

- Add a proper documentation.
- Built-in resources behaviour will not based on ``FLASK_ENV``.


0.2.0
~~~~~

Release date: 2018/7/21

-  Add three example applications.
-  ``avatars.jcrop_js()`` now default to include jQuery
   (``with_jquery=True``).
-  Built-in resources will be used when ``FLASK_ENV`` set to
   ``development``.


0.1.0
~~~~~

Release date: 2018/6/19

Initialize release.

test_flask_avatars.py
# -*- coding: utf-8 -*-
"""
    test_flask_avatars
    ~~~~~~~~~~~~~~~~~~

    :author: Grey Li <withlihui@gmail.com>
    :copyright: © 2018 Grey Li
    :license: MIT, see LICENSE for more details.
"""
import hashlib
import os
import unittest

from PIL import Image
from flask import Flask, render_template_string, current_app

from flask_avatars import Avatars, _Avatars, Identicon

basedir = os.path.abspath(os.path.dirname(__file__))


class AvatarsTestCase(unittest.TestCase):

    def setUp(self):
        self.app = Flask(__name__)

        self.app.testing = True
        self.app.secret_key = 'for test'

        self.email_hash = hashlib.md5('test@helloflask.com'.lower().encode('utf-8')).hexdigest()

        avatars = Avatars(self.app)  # noqa

        self.real_avatars = avatars

        self.avatars = _Avatars

        self.context = self.app.test_request_context()
        self.context.push()
        self.client = self.app.test_client()

    def tearDown(self):
        self.context.pop()

    def test_extension_init(self):
        self.assertIn('avatars', current_app.extensions)

    def test_gravatar(self):
        avatar_url = self.avatars.gravatar(self.email_hash)
        self.assertIn('https://gravatar.com/avatar/%s' % self.email_hash, avatar_url)
        self.assertIn('s=100', avatar_url)
        self.assertIn('r=g', avatar_url)
        self.assertIn('d=identicon', avatar_url)

        avatar_url = self.avatars.gravatar(self.email_hash, size=200, rating='x', default='monsterid',
                                           include_extension=True)
        self.assertIn('https://gravatar.com/avatar/%s' % self.email_hash, avatar_url)
        self.assertIn('s=200', avatar_url)
        self.assertIn('r=x', avatar_url)
        self.assertIn('d=monsterid', avatar_url)

    def test_robohash(self):
        avatar_url = self.avatars.robohash(self.email_hash)
        self.assertEqual(avatar_url, 'https://robohash.org/%s?size=200x200' % self.email_hash)

        avatar_url = self.avatars.robohash(self.email_hash, size=100)
        self.assertEqual(avatar_url, 'https://robohash.org/%s?size=100x100' % self.email_hash)

    def test_social_media(self):
        avatar_url = self.avatars.social_media('greyli')
        self.assertEqual(avatar_url, 'https://avatars.io/twitter/greyli/medium')

        avatar_url = self.avatars.social_media('greyli', platform='facebook', size='small')
        self.assertEqual(avatar_url, 'https://avatars.io/facebook/greyli/small')

    def test_default_avatar(self):
        avatar_url = self.avatars.default()
        self.assertEqual(avatar_url, '/avatars/static/default/default_m.jpg')

        avatar_url = self.avatars.default(size='l')
        self.assertEqual(avatar_url, '/avatars/static/default/default_l.jpg')

        avatar_url = self.avatars.default(size='s')
        self.assertEqual(avatar_url, '/avatars/static/default/default_s.jpg')

        response = self.client.get(avatar_url)
        self.assertEqual(response.status_code, 200)

    def test_load_jcrop(self):
        rv = self.avatars.jcrop_css()
        self.assertIn('<link rel="stylesheet" href="https://cdn.jsdelivr.net', rv)

        rv = self.avatars.jcrop_js()
        self.assertIn('<script src="https://cdn.jsdelivr.net', rv)
        self.assertIn('jquery.min.js', rv)

        rv = self.avatars.jcrop_js(with_jquery=False)
        self.assertIn('jquery.Jcrop.min.js', rv)
        self.assertNotIn('jquery.min.js', rv)

    def test_local_resources(self):
        current_app.config['AVATARS_SERVE_LOCAL'] = True

        response = self.client.get('/avatars/static/jcrop/js/jquery.Jcrop.min.js')
        self.assertEqual(response.status_code, 200)

        response = self.client.get('/avatars/static/jcrop/css/jquery.Jcrop.min.css')
        self.assertEqual(response.status_code, 200)

        response = self.client.get('/avatars/static/jcrop/js/jquery.min.js')
        self.assertEqual(response.status_code, 200)

        rv = self.avatars.jcrop_css()
        self.assertIn('/avatars/static/jcrop/css/jquery.Jcrop.min.css', rv)
        self.assertNotIn('<link rel="stylesheet" href="https://cdn.jsdelivr.net', rv)

        rv = self.avatars.jcrop_js()
        self.assertIn('/avatars/static/jcrop/js/jquery.Jcrop.min.js', rv)
        self.assertIn('/avatars/static/jcrop/js/jquery.min.js', rv)
        self.assertNotIn('<script src="https://cdn.jsdelivr.net', rv)

        rv = self.avatars.jcrop_js(with_jquery=False)
        self.assertIn('/avatars/static/jcrop/js/jquery.Jcrop.min.js', rv)
        self.assertNotIn('jquery.min.js', rv)

    def test_init_jcrop(self):
        rv = self.avatars.init_jcrop()
        self.assertIn('var jcrop_api,', rv)

    def test_crop_box(self):
        rv = self.avatars.crop_box()
        self.assertIn('id="crop-box"', rv)

    def test_preview_box(self):
        rv = self.avatars.preview_box()
        self.assertIn('<div id="preview-box">', rv)

    def test_render_template(self):
        rv = render_template_string('''{{ avatars.jcrop_css() }}''')
        self.assertIn('<link rel="stylesheet" href="https://cdn.jsdelivr.net', rv)

        rv = render_template_string('''{{ avatars.jcrop_js() }}''')
        self.assertIn('<script src="https://cdn.jsdelivr.net', rv)

        rv = render_template_string('''{{ avatars.init_jcrop() }}''')
        self.assertIn('var jcrop_api,', rv)

        rv = render_template_string('''{{ avatars.crop_box() }}''')
        self.assertIn('id="crop-box"', rv)

        rv = render_template_string('''{{ avatars.preview_box() }}''')
        self.assertIn('<div id="preview-box">', rv)

    def test_resize_avatar(self):
        current_app.config['AVATARS_SAVE_PATH'] = basedir
        img = Image.new(mode='RGB', size=(800, 800), color=(125, 125, 125))
        resized = self.real_avatars.resize_avatar(img, 300)
        self.assertEqual(resized.size[0], 300)

    def test_save_avatar(self):
        current_app.config['AVATARS_SAVE_PATH'] = basedir
        img = Image.new(mode='RGB', size=(800, 800), color=(125, 125, 125))
        filename = self.real_avatars.save_avatar(img)
        self.assertIn('_raw.png', filename)
        self.assertTrue(os.path.exists(os.path.join(basedir, filename)))
        os.remove(os.path.join(basedir, filename))

    def test_crop_avatar(self):
        current_app.config['AVATARS_SAVE_PATH'] = basedir
        img = Image.new(mode='RGB', size=(1000, 1000), color=(125, 125, 125))
        img.save(os.path.join(basedir, 'test.png'))
        filenames = self.real_avatars.crop_avatar('test.png', x=1, y=1, w=30, h=30)
        self.assertIn('_s.png', filenames[0])
        self.assertIn('_m.png', filenames[1])
        self.assertIn('_l.png', filenames[2])

        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[0])))
        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[1])))
        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[2])))

        file_s = Image.open(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[0]))
        self.assertEqual(file_s.size[0], current_app.config['AVATARS_SIZE_TUPLE'][0])
        file_s.close()

        for filename in filenames:
            os.remove(os.path.join(basedir, filename))

        os.remove(os.path.join(basedir, 'test.png'))

    def test_crop_default_avatar(self):
        current_app.config['AVATARS_SAVE_PATH'] = basedir
        filenames = self.real_avatars.crop_avatar(None, x=1, y=1, w=100, h=100)
        self.assertIn('_s.png', filenames[0])
        self.assertIn('_m.png', filenames[1])
        self.assertIn('_l.png', filenames[2])

        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[0])))
        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[1])))
        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[2])))

        file_s = Image.open(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[0]))
        self.assertEqual(file_s.size[0], current_app.config['AVATARS_SIZE_TUPLE'][0])
        file_s.close()

        for filename in filenames:
            os.remove(os.path.join(basedir, filename))

    def test_gravatar_mirror(self):
        mirror = self.real_avatars.gravatar(self.email_hash)
        real = self.avatars.gravatar(self.email_hash)
        self.assertEqual(mirror, real)

    def test_robohash_mirror(self):
        mirror = self.real_avatars.robohash(self.email_hash)
        real = self.avatars.robohash(self.email_hash)
        self.assertEqual(mirror, real)

    def test_social_media_mirror(self):
        mirror = self.real_avatars.social_media('grey')
        real = self.avatars.social_media('grey')
        self.assertEqual(mirror, real)

    def test_default_avatar_mirror(self):
        mirror = self.real_avatars.default()
        real = self.avatars.default()
        self.assertEqual(mirror, real)

    def test_identicon(self):
        current_app.config['AVATARS_SAVE_PATH'] = basedir

        avatar = Identicon()
        filenames = avatar.generate(text='grey')
        self.assertEqual(filenames[0], 'grey_s.png')
        self.assertEqual(filenames[1], 'grey_m.png')
        self.assertEqual(filenames[2], 'grey_l.png')

        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[0])))
        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[1])))
        self.assertTrue(os.path.exists(os.path.join(current_app.config['AVATARS_SAVE_PATH'], filenames[2])))

        # comment out these two lines to check the generated image, then delete them manually.
        for filename in filenames:
            os.remove(os.path.join(basedir, filename))


flask_avatars.py
# -*- coding: utf-8 -*-
import os
import urllib

from flask import send_from_directory, current_app, Blueprint, url_for, Markup
from randomavatar.randomavatar import Avatar


class _Avatars(object):

    @staticmethod
    def gravatar(hash, size=100, rating='g', default='identicon', include_extension=False, force_default=False):
        """
        You can get hash like this::
            import hashlib

            avatar_hash = hashlib.md5(email.lower()).hexdigest()
        :param hash
        :param size
        :param rating
        :param default

        """
        if include_extension:
            hash += '.jpg'

        default = current_app.config['AVATARS_GRAVATAR_DEFAULT'] or default
        query_string = urllib.urlencode({'s': int(size), 'r': rating, 'd': default})

        if force_default:
            query_string += '&q=y'
        return 'https://gravatar.com/avatar/' + hash + '?' + query_string

    @staticmethod
    def robohash(text, size=200):
        return 'https://robohash.org/{text}?size={size}x{size}'.format(text=text, size=size)

    @staticmethod
    def social_media(username, platform='twitter', size='medium'):
        """
        facebook, instagram, twitter, gravatar
        """
        return 'https://avatars.io/{platform}/{username}/{size}'.format(platform=platform, username=username, size=size)

    @staticmethod
    def default(size='m'):
        return url_for('static', filename='default/{size}.jpg'.format(size=size))


    @staticmethod
    def load_jcrop():

        preview_size = 180
        crop_size = 500
        return Markup('''
        
<style>
#preview-pane .preview-container {
  width: %dpx;
  height: %dpx;
  overflow: hidden;
}

#crop-box {
  max-width: %dpx;
  display: block;
}
        </style>
        ''' % (preview_size, preview_size, crop_size))

    @staticmethod
    def crop_box(url):
        if url is None:
            url = '/avatar/%s'
        return '<img src="%s" id="cropbox">' % url

    @staticmethod
    def preview_pane(url):
        return '''
        <div id="preview-pane">
        <div class="preview-container">
          <img src="%s"
               class="jcrop-preview" alt="Preview"/>
        </div>
      </div>''' % url




class Identicon(object):
    def __init__(self, rows=10, cols=10):
        self.rows = rows
        self.cols = cols

    def generate(self, text, path, id, size=100):
        avatar = Avatar(rows=self.rows, columns=self.cols)
        image_byte_array = avatar.get_image(
            string=str(text),
            width=int(size),
            height=int(size),
            pad=int(size * 0.05))
        avatar.save(image_byte_array,
                    save_location=os.path.join(path, '%s_%d.png' % (id, size)))



class Avatars(object):
    def __init__(self, app=None):
        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        if not hasattr(app, 'extensions'):
            app.extensions = {}
        app.extensions['avatars'] = _Avatars
        app.context_processor(self.context_processor)

        blueprint = Blueprint('avatars', __name__)
        app.register_blueprint(blueprint)

        # settings
        app.config.setdefault('AVATARS_GRAVATAR_DEFAULT', 'identicon')
        app.config.setdefault('AVATARS_UPLOAD_PATH', None)

        @blueprint.route('/avatar/<int:id>/<int:size>')
        def avatar(id, size):
            path = current_app.config['AVATARS_UPLOAD_PATH']
            filename = '%s_%d.png' % (id, size)
            return send_from_directory(path, filename)


    @staticmethod
    def context_processor():
        return {
            'avatars': current_app.extensions['avatars']
        }


examples/identicon/app.py
# -*- coding: utf-8 -*-
"""
    identicon.app
    ~~~~~~~~~~~~~
    Example application that demonstrate how to generate identicon.

    :author: Grey Li <withlihui@gmail.com>
    :copyright: © 2018 Grey Li
    :license: MIT, see LICENSE for more details.
"""
import os
import uuid

from flask import Flask, render_template, url_for, send_from_directory
from flask_avatars import Avatars, Identicon

basedir = os.path.abspath(os.path.dirname(__name__))

app = Flask(__name__)
app.config['AVATARS_SAVE_PATH'] = os.path.join(basedir, 'avatars')

avatars = Avatars(app)


# serve avatar image
@app.route('/avatars/<path:filename>')
def get_avatar(filename):
    return send_from_directory(app.config['AVATARS_SAVE_PATH'], filename)


@app.route('/')
def index():
    avatar = Identicon()
    random_text = uuid.uuid4().hex
    filenames = avatar.generate(text=random_text)
    url_s = url_for('get_avatar', filename=filenames[0])
    url_m = url_for('get_avatar', filename=filenames[1])
    url_l = url_for('get_avatar', filename=filenames[2])
    return render_template('index.html', url_s=url_s, url_m=url_m, url_l=url_l)


flask_avatars/__init__.py
# -*- coding: utf-8 -*-
"""
    flask_avatars
    ~~~~~~~~~~~~~

    :author: Grey Li <withlihui@gmail.com>
    :copyright: © 2018 Grey Li
    :license: MIT, see LICENSE for more details.
"""
import os
try:
    from urllib.parse import urlencode
except ImportError:
    from urllib import urlencode

from uuid import uuid4

from PIL import Image
from flask import current_app, Blueprint, url_for
from markupsafe import Markup
from .identicon import Identicon  # noqa


class _Avatars(object):

    @staticmethod
    def gravatar(hash, size=100, rating='g', default='identicon', include_extension=False, force_default=False):
        """Pass email hash, return Gravatar URL. You can get email hash like this::

            import hashlib
            avatar_hash = hashlib.md5(email.lower().encode('utf-8')).hexdigest()

        Visit https://en.gravatar.com/site/implement/images/ for more information.

        :param hash: The email hash used to generate avatar URL.
        :param size: The size of the avatar, default to 100 pixel.
        :param rating: The rating of the avatar, default to ``g``
        :param default: The type of default avatar, default to ``identicon``.
        :param include_extension: Append a '.jpg' extension at the end of URL, default to ``False``.
        :param force_default: Force to use default avatar, default to ``False``.
        """
        if include_extension:
            hash += '.jpg'

        default = default or current_app.config['AVATARS_GRAVATAR_DEFAULT']
        query_string = urlencode({'s': int(size), 'r': rating, 'd': default})

        if force_default:
            query_string += '&q=y'
        return 'https://gravatar.com/avatar/' + hash + '?' + query_string

    @staticmethod
    def robohash(text, size=200):
        """Pass text, return Robohash-style avatar (robot).
        Visit https://robohash.org/ for more information.

        :param text: The text used to generate avatar.
        :param size: The size of the avatar, default to 200 pixel.
        """
        return 'https://robohash.org/{text}?size={size}x{size}'.format(text=text, size=size)

    @staticmethod
    def social_media(username, platform='twitter', size='medium'):
        """Return avatar URL at social media.
        Visit https://avatars.io for more information.

        :param username: The username of the social media.
        :param platform: One of facebook, instagram, twitter, gravatar.
        :param size: The size of avatar, one of small, medium and large.
        """
        return 'https://avatars.io/{platform}/{username}/{size}'.format(
            platform=platform, username=username, size=size)

    @staticmethod
    def default(size='m'):
        """Return built-in default avatar.

        :param size: The size of avatar, one of s, m, l.
        :return: Default avatar URL
        """
        return url_for('avatars.static', filename='default/default_{size}.jpg'.format(size=size))

    @staticmethod
    def jcrop_css(css_url=None):
        """Load jcrop css file.

        :param css_url: The custom CSS URL.
        """
        if css_url is None:
            if current_app.config['AVATARS_SERVE_LOCAL']:
                css_url = url_for('avatars.static', filename='jcrop/css/jquery.Jcrop.min.css')
            else:
                css_url = 'https://cdn.jsdelivr.net/npm/jcrop-0.9.12@0.9.12/css/jquery.Jcrop.min.css'
        return Markup('<link rel="stylesheet" href="%s">' % css_url)

    @staticmethod
    def jcrop_js(js_url=None, with_jquery=True):
        """Load jcrop Javascript file.

        :param js_url: The custom JavaScript URL.
        :param with_jquery: Include jQuery or not, default to ``True``.
        """
        serve_local = current_app.config['AVATARS_SERVE_LOCAL']

        if js_url is None:
            if serve_local:
                js_url = url_for('avatars.static', filename='jcrop/js/jquery.Jcrop.min.js')
            else:
                js_url = 'https://cdn.jsdelivr.net/npm/jcrop-0.9.12@0.9.12/js/jquery.Jcrop.min.js'

        if with_jquery:
            if serve_local:
                jquery = '<script src="%s"></script>' % url_for('avatars.static', filename='jcrop/js/jquery.min.js')
            else:
                jquery = '<script src="https://cdn.jsdelivr.net/npm/jcrop-0.9.12@0.9.12/js/jquery.min.js"></script>'
        else:
            jquery = ''
        return Markup('''%s\n<script src="%s"></script>
        ''' % (jquery, js_url))

    @staticmethod
    def crop_box(endpoint=None, filename=None):
        """Create a crop box.

        :param endpoint: The endpoint of view function that serve avatar image file.
        :param filename: The filename of the image that need to be crop.
        """
        crop_size = current_app.config['AVATARS_CROP_BASE_WIDTH']

        if endpoint is None or filename is None:
            url = url_for('avatars.static', filename='default/default_l.jpg')
        else:
            url = url_for(endpoint, filename=filename)
        return Markup('<img src="%s" id="crop-box" style="max-width: %dpx; display: block;">' % (url, crop_size))

    @staticmethod
    def preview_box(endpoint=None, filename=None):
        """Create a preview box.

        :param endpoint: The endpoint of view function that serve avatar image file.
        :param filename: The filename of the image that need to be crop.
        """
        preview_size = current_app.config['AVATARS_CROP_PREVIEW_SIZE'] or current_app.config['AVATARS_SIZE_TUPLE'][2]

        if endpoint is None or filename is None:
            url = url_for('avatars.static', filename='default/default_l.jpg')
        else:
            url = url_for(endpoint, filename=filename)
        return Markup('''
        <div id="preview-box">
        <div class="preview-box" style="width: %dpx; height: %dpx; overflow: hidden;">
          <img src="%s" class="jcrop-preview" alt="Preview"/>
        </div>
      </div>''' % (preview_size, preview_size, url))

    @staticmethod
    def init_jcrop(min_size=None):
        """Initialize jcrop.

        :param min_size: The minimal size of crop area.
        """
        init_x = current_app.config['AVATARS_CROP_INIT_POS'][0]
        init_y = current_app.config['AVATARS_CROP_INIT_POS'][1]
        init_size = current_app.config['AVATARS_CROP_INIT_SIZE'] or current_app.config['AVATARS_SIZE_TUPLE'][2]

        if current_app.config['AVATARS_CROP_MIN_SIZE']:
            min_size = min_size or current_app.config['AVATARS_SIZE_TUPLE'][2]
            min_size_js = 'jcrop_api.setOptions({minSize: [%d, %d]});' % (min_size, min_size)
        else:
            min_size_js = ''
        return Markup('''
<script type="text/javascript">
    jQuery(function ($) {
      // Create variables (in this scope) to hold the API and image size
      var jcrop_api,
          boundx,
          boundy,

          // Grab some information about the preview pane
          $preview = $('#preview-box'),
          $pcnt = $('#preview-box .preview-box'),
          $pimg = $('#preview-box .preview-box img'),

          xsize = $pcnt.width(),
          ysize = $pcnt.height();

      $('#crop-box').Jcrop({
        onChange: updatePreview,
        onSelect: updateCoords,
        setSelect: [%s, %s, %s, %s],
        aspectRatio: 1
      }, function () {
        // Use the API to get the real image size
        var bounds = this.getBounds();
        boundx = bounds[0];
        boundy = bounds[1];
        // Store the API in the jcrop_api variable
        jcrop_api = this;
        %s
        jcrop_api.focus();
        // Move the preview into the jcrop container for css positioning
        $preview.appendTo(jcrop_api.ui.holder);
      });

      function updatePreview(c) {
        if (parseInt(c.w) > 0) {
          var rx = xsize / c.w;
          var ry = ysize / c.h;
          $pimg.css({
            width: Math.round(rx * boundx) + 'px',
            height: Math.round(ry * boundy) + 'px',
            marginLeft: '-' + Math.round(rx * c.x) + 'px',
            marginTop: '-' + Math.round(ry * c.y) + 'px'
          });
        }
      }
    });

    function updateCoords(c) {
      $('#x').val(c.x);
      $('#y').val(c.y);
      $('#w').val(c.w);
      $('#h').val(c.h);
    }
  </script>
            ''' % (init_x, init_y, init_size, init_size, min_size_js))


class Avatars(object):
    def __init__(self, app=None):
        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        if not hasattr(app, 'extensions'):
            app.extensions = {}
        app.extensions['avatars'] = _Avatars
        app.context_processor(self.context_processor)

        blueprint = Blueprint('avatars', __name__,
                              static_folder='static',
                              static_url_path='/avatars' + app.static_url_path)
        app.register_blueprint(blueprint)

        self.root_path = blueprint.root_path

        # TODO: custom file extension support
        # settings
        app.config.setdefault('AVATARS_GRAVATAR_DEFAULT', 'identicon')

        app.config.setdefault('AVATARS_SERVE_LOCAL', False)

        app.config.setdefault('AVATARS_SAVE_PATH', None)
        app.config.setdefault('AVATARS_SIZE_TUPLE', (30, 60, 150))
        # Identicon
        app.config.setdefault('AVATARS_IDENTICON_COLS', 7)
        app.config.setdefault('AVATARS_IDENTICON_ROWS', 7)
        app.config.setdefault('AVATARS_IDENTICON_BG', None)
        # Jcrop
        app.config.setdefault('AVATARS_CROP_BASE_WIDTH', 500)
        app.config.setdefault('AVATARS_CROP_INIT_POS', (0, 0))
        app.config.setdefault('AVATARS_CROP_INIT_SIZE', None)
        app.config.setdefault('AVATARS_CROP_PREVIEW_SIZE', None)
        app.config.setdefault('AVATARS_CROP_MIN_SIZE', None)

        # @blueprint.route('/%s/<path:filename>/<size>' % app.config['AVATARS_STATIC_PREFIX'])
        # def static(filename_m):
        #     path = current_app.config['AVATARS_SAVE_PATH']
        #     filename = '%s_%s.png' % (filename, size)
        #     return send_from_directory(path, filename)

    @staticmethod
    def context_processor():
        return {
            'avatars': current_app.extensions['avatars']
        }

    def resize_avatar(self, img, base_width):
        """Resize an avatar.

        :param img: The image that needs to be resize.
        :param base_width: The width of output image.
        """
        w_percent = (base_width / float(img.size[0]))
        h_size = int((float(img.size[1]) * float(w_percent)))
        img = img.resize((base_width, h_size), Image.BICUBIC)
        return img

    def save_avatar(self, image):
        """Save an avatar as raw image, return new filename.

        :param image: The image that needs to be saved.
        """
        path = current_app.config['AVATARS_SAVE_PATH']
        filename = uuid4().hex + '_raw.png'
        image.save(os.path.join(path, filename))
        return filename

    def crop_avatar(self, filename, x, y, w, h, uuid_filename = True):
        """Crop avatar with given size, return a list of file name: [filename_s, filename_m, filename_l].

        :param filename: The raw image's filename.
        :param x: The x-pos to start crop.
        :param y: The y-pos to start crop.
        :param w: The crop width.
        :param h: The crop height.
        """
        x = int(x)
        y = int(y)
        w = int(w)
        h = int(h)

        sizes = current_app.config['AVATARS_SIZE_TUPLE']

        if not filename:
            path = os.path.join(self.root_path, 'static/default/default_l.jpg')
        else:
            path = os.path.join(current_app.config['AVATARS_SAVE_PATH'], filename)

        raw_img = Image.open(path)

        base_width = current_app.config['AVATARS_CROP_BASE_WIDTH']

        if raw_img.size[0] >= base_width:
            raw_img = self.resize_avatar(raw_img, base_width=base_width)

        cropped_img = raw_img.crop((x, y, x + w, y + h))

        if uuid_filename:
            filename = uuid4().hex

        avatar_s = self.resize_avatar(cropped_img, base_width=sizes[0])
        avatar_m = self.resize_avatar(cropped_img, base_width=sizes[1])
        avatar_l = self.resize_avatar(cropped_img, base_width=sizes[2])

        filename_s = filename + '_s.png'
        filename_m = filename + '_m.png'
        filename_l = filename + '_l.png'

        path_s = os.path.join(current_app.config['AVATARS_SAVE_PATH'], filename_s)
        path_m = os.path.join(current_app.config['AVATARS_SAVE_PATH'], filename_m)
        path_l = os.path.join(current_app.config['AVATARS_SAVE_PATH'], filename_l)

        avatar_s.save(path_s, optimize=True, quality=85)
        avatar_m.save(path_m, optimize=True, quality=85)
        avatar_l.save(path_l, optimize=True, quality=85)

        return [filename_s, filename_m, filename_l]

    @staticmethod
    def gravatar(*args, **kwargs):
        return _Avatars.gravatar(*args, **kwargs)

    @staticmethod
    def robohash(*args, **kwargs):
        return _Avatars.robohash(*args, **kwargs)

    @staticmethod
    def social_media(*args, **kwargs):
        return _Avatars.social_media(*args, **kwargs)

    @staticmethod
    def default(*args, **kwargs):
        return _Avatars.default(*args, **kwargs)
